components/common/src/error.rs

```diff
@@ -55,17 +56,76 @@ impl fmt::Display for CommandExecutionError {
     }
 }
 
+#[derive(Debug)]
+pub enum APIFailure {
+    DownloadFailed(String),
+    UploadFailed(String),
+    Package(FullyQualifiedPackageIdent, PackageTarget),
+    OriginKeyRev(String, String),
+    LatestOriginKey(String),
+}
```

Comment:
I love breaking out these errors in a more granular way like this, but I think there are some improvements we can make to better preserve encapsulation and also make it even easier to manage these errors.

First, as API errors, it feels like this particular enum really belongs in the API client crate. I'd also argue that it should actually implement `Error` directly.

One neat thing about enums is that the individual instances can serve as their own constructors. That is, if `DownloadFailed` were defined as `DownloadFailed(usize, FullyQualifiedPackageIdent, PackageTarget, Error)`, then you wouldn't actually need the `APIFailure::package_download_failed` function; you could just call, say, `DownloadFailed(5, my_ident, my_target, e)` and call it a day. What _package_download_failed_  is doing is actually what `impl std::fmt::Display for APIFailure` should be doing.

What the `Error::BuilderAPITransferError` enum case is actually doing is providing a mapping of the API client error type into the `habitat_common::error::Error` type. This is often done with `From` implementations for errors, which is what the `?` operator uses to coerce errors on your behalf. For instance, you'll often see a bunch of implementations in our various error modules like this:

```rust
impl From<io::Error> for Error {
    fn from(err: io::Error) -> Self { Error::IO(err) }
}
```
That's basically taking a error from some other crate we depend on (here, a standard library I/O error) and mapping it to one of _our_ error instances, which ensures that everything coming out of _our_ crate has a unified error type. That, in turn, makes it easier for consumers of _our_ crate to deal with errors that come from it.

With this implementation in place, we really don't have to worry about explicitly mapping standard library I/O errors into our own errors; any time we have a function that calls a standard library function that could return one of those, simply adding a `?` to that call will automatically invoke this `From` implementation to coerce it into our own error type, free of charge.

This _can_ also end up being a bad thing if it is used indiscriminately; you may not actually want every conceivable I/O error to map to the same error case in your crate. Perhaps problems reading a configuration file should actually be a case of a `ConfigurationError`, for example, which is more meaningful and carries more useful information that "I tried to do an I/O operation somewhere and it failed; sorry! :shrug: " We actually have a lot of this in the codebase right now, to our detriment. However, I think a big reason this is a problem for us is that we tend to have giant catch-all error cases at the top level of our crates that effectively flatten out our error mapping in this bad way. If you had a `ConfigurationError` that was thrown by anything in your `config.rs` module (and submodules!), then you could map all your I/O errors there to `ConfigurationErrors`, and then at the top level of your crate, you could map `ConfigurationError` to whatever your top-level `Error` was. It's a matter of capturing whatever content is important at the site of the error, and then providing the means to bubble that up as appropriate.

I did some of this in the container exporter a little while ago, where I have an `Engine` abstraction that has its own error type, and then Docker and Buildah implementations of that `Engine` that have their _own_ error types, each of which get progressively bubbled up to higher-and-higher levels. (e.g., https://github.com/habitat-sh/habitat/blob/9224683d30f5cc7c276ac179ae0f3d9b8eee02d6/components/pkg-export-container/src/engine/buildah.rs#L33-L41). This helps to keep knowledge localized, but also provides clear paths for information to flow from one place to another without any degradation.

(Just to be clear, I'm absolutely _not_ advising that every single module in a crate necessarily must have their own error type, and that their parent module then has to coerce all errors from them and their sibling modules into a higher-level error type, and on and on, until you have a crate that has 52 different error types. Rather, I'm arguing that distinct logical subdomains within a crate could very well benefit from having their own error type that adequately captures all the necessary nuance of the error conditions that they can produce. This is, in fact, _exactly_ why `std::io::Error` exists in the first place.) 

Viewed from this perspective, let's take a look at this code, picked arbitrarily from `hab/src/command/origin/key/upload.rs`
```rust
 return Err(CommonError::BuilderAPITransferError(APIFailure::key_upload_failed(API_RETRIES, &name, &rev, e.into())).into());
```
Here, we're in the `hab` CLI crate, but we are manually creating an error that is logically from the API crate (using `APIFailure::key_upload_failed`), manually wrapping that in an error from _another_ crate that isn't even in this code flow (`habitat_common`, via `CommonError::BuilderAPITransferError`), and then finally converting that into an instance of the catch-all error case from _our_ crate (using `into()`, which invokes the code from a `From` implementation: https://github.com/habitat-sh/habitat/blob/9224683d30f5cc7c276ac179ae0f3d9b8eee02d6/components/common/src/error.rs#L252-L254).

We're concerning ourselves with too many low-level concerns in what is fairly high-level code here. The root of the error here, the `APIFailure` should be something that is produced completely from within the API crate itself, because that is where all knowledge about that error is located. If we had some additional context up here in the CLI crate that is not present in the API crate, but is nevertheless important to capture, then that would signal that our CLI crate should have some kind of error instance that wraps an API error, but also adds whatever important information it needs to. We shouldn't be creating a `habitat_common` error in the CLI crate, though, because only the `habitat_common` crate should be creating its errors.

If things were set up like I'm arguing for, then we could remove that entire `return` clause and simply replace it with `?`, and everything would just work. Additionally, it would declutter the code so you could focus on what it's actually doing, rather than wading through lots of error handling. We'd basically be taking a lot of repeated logic that lives in `match` and `if` statements, and encoding them into the structure of our type system.

Reply:
Completely agree! This seems like the "correct" way to do error handling. Two crates that may be useful are [`thiserror`](https://github.com/dtolnay/thiserror) and [`anyhow`](https://github.com/dtolnay/anyhow). See [here](https://github.com/dtolnay/thiserror#comparison-to-anyhow) for a good explanation of when to use each one. I think `thiserror` will be especially useful in cutting down the boilerplate of having many error types. The great thing about `thiserror` is it solely deals with `std::error::Error` (vs `failure` which introduces new traits).

Reply:
@davidMcneil Ooh, those both look incredibly useful. I'm all up for some experiments with them :+1:

Reply:
@christophermaier Thanks for the detailed review in this section. After reading carefully, I believe I've implemented the patterns you're suggesting in https://github.com/habitat-sh/habitat/pull/7874/commits/26704a92a178597076d034d64aa498acfd33132d